cms/static_src/javascript/auth.js

```diff
@@ -0,0 +1,149 @@
+  const csrfToken = getCookieByName(csrfCookieName);
+  if (!csrfToken) {
+    console.error('[WAGTAIL] CSRF token not found.');
+    return null;
+  }
+
+  const headers = {
+    'Content-Type': 'application/json',
+    [csrfHeaderName]: csrfToken,
+    ...additionalHeaders,
+  };
+
+  const config = {
+    method,
+    headers,
+    credentials: 'include',
+    ...(body && { body: JSON.stringify(body) }),
+  };
+
+  try {
+    return await fetch(url, config);
+  } catch (error) {
+    console.error(`[WAGTAIL] Error during ${method} request to ${url}:`, error);
+    return null;
+  }
+};
+
+// Initiates the logout process and handles redirection.
+const logout = async () => {
+  console.info('[WAGTAIL] Initiating logout...');
+  try {
+    const response = await fetchWithCsrf(logoutURL);
+    if (response?.ok) {
+      const redirectURL = response.redirected ? response.url : origin;
+      console.info('[WAGTAIL] Logout successful. Redirecting to:', redirectURL);
+      window.location.href = redirectURL;
+    } else {
+      console.error('[WAGTAIL] Logout failed:', response);
+      // Reload the page to handle any session inconsistencies.
+      window.location.reload();
+    }
+  } catch (error) {
+    console.error('[WAGTAIL] Exception during logout:', error);
+  }
+};
+
+// Session configuration
+const sessionConfig = {
+  timeOffsets: { passiveRenewal: parseInt(sessionRenewalOffsetSeconds, 10) * 1000 }, // milliseconds
+  apiEndpoints: {
+    renewSession: authTokenRefreshUrl,
+  },
+  onRenewSuccess: async (sessionExpiryTime, refreshExpiryTime) => {
+    console.debug(
+      `[WAGTAIL] Session renewed! (Session: ${sessionExpiryTime}, Refresh: ${refreshExpiryTime})`,
+    );
+    try {
+      const extendResponse = await fetchWithCsrf(extendSessionURL);
+      if (extendResponse?.ok) {
+        console.info('[WAGTAIL] Wagtail session extended successfully.');
+      } else {
+        console.error('[WAGTAIL] Session extension failed:', extendResponse);
+        await logout();
+      }
+    } catch (error) {
+      console.error('[WAGTAIL] Error extending session:', error);
+      await logout();
+    }
+  },
+  onRenewFailure: async (error) => {
+    console.error('[WAGTAIL] Session renewal failed:', error);
+    await logout();
+  },
+  onSessionValid: (sessionExpiryTime, refreshExpiryTime) => {
+    console.debug(
+      `[WAGTAIL] Session valid. (Session: ${sessionExpiryTime}, Refresh: ${refreshExpiryTime})`,
+    );
+  },
+  onSessionInvalid: async () => {
+    console.warn('[WAGTAIL] Session invalid.');
+    await logout();
+  },
+  onError: async (error) => {
+    console.error('[WAGTAIL] General error:', error);
+    await logout();
+  },
+};
+
+// Initialise SessionManagement only if the user is logged in via Cognito (indicated by the presence of an id_token cookie).
+// Don't load if it's in an iframe (e.g. the Wagtail admin interface preview).
+document.addEventListener('DOMContentLoaded', () => {
```

Comment:
Question: On line 18 we log an error if we couldn't retrieve the config. However we continue nonetheless. Should we add this event listener if and only if line 16 was successful, i.e. `authConfig` isn't still an empty object?

Reply:
Look like a sensible change.